File: packages/contracts/src/shop.ts
import { z } from 'zod';

/**
 * Shopping feed (Wave F2) — GET /shop.
 *
 * A page of ranked, mixed cards. Every card carries the fields the client needs for event
 * capture: `feedPosition`, `cardType`, and the `exploration` flag (echoed into the
 * impression event so exploration engagement is measurable apart from exploited positions).
 *
 * MONETIZATION BOUNDARY: cards expose a `productId`, NEVER an outbound/affiliate URL. To
 * open a product the client mints a click via POST /clicks and follows GET /out/{clickId}.
 */

/** Wire-safe product fields (no outbound URL — click via productId). */
export const feedProductSchema = z.object({
  productId: z.string().uuid(),
  name: z.string(),
  brand: z.string().nullable().optional(),
  merchant: z.string().nullable().optional(),
  imageUrl: z.string().nullable().optional(),
  price: z.number().nullable().optional(),
  currency: z.string().nullable().optional(),
  category: z.string().nullable().optional(),
});
export type FeedProduct = z.infer<typeof feedProductSchema>;

/** Fields shared by every card, consumed by the client's event emitter. */
const cardCommon = {
  feedPosition: z.number().int(),
  cardType: z.enum(['product', 'outfit']),
  exploration: z.boolean(),
  score: z.number().optional(),
};

/** ~70% of the feed: a buyable product + how many outfits it unlocks. */
export const productCardSchema = z.object({
  type: z.literal('product'),
  product: feedProductSchema,
  unlockCount: z.number().int(),
  headline: z.string(),
  gapContext: z
    .object({
      fillsEmptyOccasion: z.boolean().optional(),
      category: z.string().nullable().optional(),
    })
    .partial()
    .optional(),
  ...cardCommon,
});
export type ProductCard = z.infer<typeof productCardSchema>;

/** ~30% of the feed: owned items + 1 buyable that completes the look. */
export const outfitCardSchema = z.object({
  type: z.literal('outfit'),
  occasion: z.string().nullable().optional(),
  ownedItemIds: z.array(z.string().uuid()),
  buyable: feedProductSchema,
  buyableProductId: z.string().uuid(),
  unlockCount: z.number().int(),
  collageUrl: z.string().nullable().optional(),
  rationale: z.string(),
  ...cardCommon,
});
export type OutfitCard = z.infer<typeof outfitCardSchema>;

export const shopCardSchema = z.discriminatedUnion('type', [
  productCardSchema,
  outfitCardSchema,
]);
export type ShopCard = z.infer<typeof shopCardSchema>;

export const shopFeedResponseSchema = z.object({
  cards: z.array(shopCardSchema),
  cursor: z.number().int(),
  sessionId: z.string(),
  hasMore: z.boolean(),
  framing: z.enum(['personalized', 'starter_looks']),
  diagnostics: z.record(z.string(), z.unknown()).optional(),
});
export type ShopFeedResponse = z.infer<typeof shopFeedResponseSchema>;

File: app/api/routes/shop.py
"""GET /shop — the Stage-1 shopping feed (Wave F2).

Returns a page of ranked, mixed cards for the authenticated user: ~70% product cards
("unlocks N outfits" + gap preview) and ~30% outfit cards (owned items + 1 buyable, via the
composer + a PIL collage). Ranked by app.ranking (taste + wardrobe-gap + price_fit +
quality − fatigue), re-ranked for diversity/calibration with an exploration slice, and
paginated by a session watermark so pages of one browse stay consistent.

Feed serve is $0 API (pgvector + numpy + the pure assembler — no model call).

MONETIZATION BOUNDARY: cards carry a productId, never an outbound/affiliate URL. To open a
product the client mints a click via POST /clicks and follows GET /out/{click_id}
(app/monetization). This route imports nothing from app.monetization.
"""
from __future__ import annotations

import logging
import uuid
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.core.config import settings
from app.dependencies import get_current_user, get_db
from app.models import User
from app.ranking.feed import build_feed

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/shop", tags=["shop"])


class ShopFeedResponse(BaseModel):
    """A page of the feed. ``cards`` is a heterogeneous list (product | outfit); each card
    carries feedPosition / cardType / exploration for the client's event capture. ``sessionId``
    is the watermark the client MUST echo on the next page for a stable feed."""

    cards: List[Dict[str, Any]] = Field(default_factory=list)
    cursor: int
    sessionId: str
    hasMore: bool
    framing: str  # "personalized" | "starter_looks"
    diagnostics: Dict[str, Any] = Field(default_factory=dict)


@router.get("", response_model=ShopFeedResponse)
def get_shop_feed(
    cursor: int = Query(0, ge=0, description="Offset of the page to return (0 = first)."),
    sessionId: Optional[str] = Query(
        None, description="Session watermark from a prior page; omit on the first request."
    ),
    pageSize: Optional[int] = Query(
        None, ge=1, le=100, description="Cards per page (defaults to the configured page size)."
    ),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> ShopFeedResponse:
    # New browse → mint a watermark (randomness here is fine; the ranker itself is seeded
    # deterministically FROM this value so pagination is reproducible).
    session_id = sessionId or uuid.uuid4().hex

    page = build_feed(
        db,
        current_user.id,
        session_id=session_id,
        cursor=cursor,
        page_size=pageSize,
    )
    return ShopFeedResponse(
        cards=page.cards,
        cursor=page.cursor,
        sessionId=page.session_id,
        hasMore=page.has_more,
        framing=page.framing,
        diagnostics=page.diagnostics,
    )
